bpf:nat: Fix DSR support for remote NodePort services

gyutaeb opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

---

<!-- Description of change -->

# Problem Statement

- When DSR and PER_PACKET_LB are enabled, connections fail if a client pod sends a request to a remote nodes's NodePort service while the server pod is located on the same node as the client. 
- This issue causes the Cilium connectivity tests to fail.
- For more details, refer to #41962 

# Root Cause

<img width="8284" height="2684" alt="image" src="https://github.com/user-attachments/assets/e28423c8-ba16-4bf6-9eb8-0733ef0fc1fd" />

- The root cause is that remote node only performs DNAT. This sets the packets's source address to client's node address, leading to a hairpin problem. Consequently, the originating node cannot perform the necessary REV NAT for the reply packets.
- The problem occurs through the following path:
  - Request
    - Client Pod -> Remote Node:nodePort
    - Local Node:sport -> Remote Node:nodePort (masquerading)
    - Local Node:sport -> Backend Pod:port (DSR DNAT)
    - Send the above packet over Geneve From Remote Node
    - Local Node route above packet to Backend Pod:port
   - Reply
     - BackendPod:port -> Local Node:sport
     - Route to loopback without REVNAT. Because of destination address is Local Node

# Solution Summary

<img width="4664" height="2604" alt="image" src="https://github.com/user-attachments/assets/0b254aa2-e33c-4d32-8521-fbadbf21a795" />

- To resolve this, remote service requests are now handled on the source node when DSR is enabled, similar to the behavior of socket-level load balancing.

# Key Consideration

- If we choose to address this by egressing to a remote node, a REV NAT stage corresponding to the native device masquerading is required. Because REV NAT executes only on the native devices's ingress path, we conclude it is better to handle this on the local node--analogous to the BPF socket load-balancer. Performing REV NAT in cil_from_container would unnecessarily increase complexity.

# Release Note
```release-note
Fix in-cluster NodePort connectivity failure in DSR mode when SocketLB
is disabled. When a pod accesses a NodePort service via a remote node's
IP (instead of the ClusterIP) and the selected backend resides on the
same node as the client, the connection fails due to missing reverse NAT
on the reply path.
```

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -139,6 +178,14 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+			if (IS_ERR(ret))
+				return ret;
+		}
+#endif /* ENABLE_DSR */
```
I don't understand why you're doing special things here for DSR. Once we wildcard-matched the Nodeport service, isn't the `svc`'s `rev_nat_index` (which we stash into the CT entry) sufficient to later perform the RevDNAT operation for replies?

gyutaeb replied:
Thanks for the question! @julianwiedmann 

The `rev_nat_index` from the wildcard match isn't enough here. Because of the wildcard-matching is against `0.0.0.0:nodeport`, and the remote node's Nodeport IP is not present in the local service list. The backend's reply is just `backend -> client`, but we must RevNAT it to `remote-node-IP:<nodeport> -> client`. Since that remote node IP isn't in the local service list, we pre-create an SNAT mapping (same mechanism as DSR) so the reply path can be RevNAT.

WIthout an SNAT mapping entry, the reply end up RevDNATed as `0.0.0.0 -> client pod`, which is incorrect.
I initially tried to avoid using an SNAT mapping entry, but given the current design this is the most reliable approach.

SocketLB doesn't need this mechanism because the wildcard match happens at the socket layer. Socket's destination changed to backend address, so no additinal SNAT mapping is required for SocketLB.

julianwiedmann replied:
> The rev_nat_index from the wildcard match isn't enough here. Because of the wildcard-matching is against 0.0.0.0:nodeport, and the remote node's Nodeport IP is not present in the local service list. The backend's reply is just backend -> client, but we must RevNAT it to remote-node-IP:<nodeport> -> client. Since that remote node IP isn't in the local service list, we pre-create an SNAT mapping (same mechanism as DSR) so the reply path can be RevNAT.

uff of course, thanks. We can obtain the original port from the service entry, but not the specific NodeIP. How annoying. I'd really want to get rid of the NAT entries for DSR :/.

edit: have you already looked at whether the GC for NAT entries works-as-is for this usage? I remember it expects very specific CT entries to keep the NAT entries alive.

> SocketLB doesn't need this mechanism because the wildcard match happens at the socket layer. Socket's destination changed to backend address, so no additinal SNAT mapping is required for SocketLB.

I think the SocketLB actually still manages a RevNAT map somewhere.

gyutaeb replied:
@julianwiedmann @aditighag 
Thanks for your patience.

> edit: have you already looked at whether the GC for NAT entries works-as-is for this usage? I remember it expects very specific CT entries to keep the NAT entries alive.

GC for NAT entries is working correctly. 
In the packet flow this PR addresses, the NAT entry's key is created as a `backend -> client pod` tuple. This entry is properly garbage-collected by the existing ["GC NAT entries created by DSR"](https://github.com/cilium/cilium/pull/9473/commits/107d8f9f7767434142be9ccf29e3b6b7cc0530be) logic that was added when DSR was introduced. Because the tuple generated by `snat_v4_create_dsr()` uses the same `backend -> client pod` key.
I also confirmed that the NAT entries are actually deleted when the GC runs.

> I think the SocketLB actually still manages a RevNAT map somewhere.

SocketLB keeps a RevNAT map, but that is mainly to support userspace syscalls such as `recvmsg()/ getpeername()`. In the `client pod -> remote-node:<nodeport>` case, userspace sees it as if it taking to `remote-node:<nodeport>`. However, when you look at the actual socket with `ss/netstat`, the socket is created as `client pod <-> endpoint`. So we don't need additional DSR-specific NAT entries in this path for SocketLB.

aditighag reviewed (commented):
Thanks for your contribution, and the detailed PR description.
IIUC, there are two main changes in the PR: (1) Perform LB wildcard svc lookup and xlate for nodeport-vip:port at the source node for DSR. (2) Create CT mappings for the xlate'd traffic so that it can be reverse xlate'd.

A high-level question: Is there a reason why (1) is implemented only for DSR? Why not align it with the socket-LB implementation for all cases?

I’d also suggest extracting the service wildcard lookup code into a helper function that explicitly highlights the functionality (similar to socket-LB). This would make it easier to draw parallels with the socket-LB implementation for feature parity.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -1931,6 +2002,28 @@ ipv4_policy(struct __ctx_buff *ctx, struct iphdr *ip4, __u32 src_label,
+					if (IS_ERR(ret))
+						return ret;
+					break;
+				}
```
I think we need to do better here, and not have two near-identical calls to `lb4_rev_nat()`.

gyutaeb replied:
https://github.com/cilium/cilium/pull/41963#discussion_r2583427880

julianwiedmann reviewed (commented):
I'd want to see if we can land https://github.com/cilium/cilium/pull/43017 first, and how that changes the picture. My strong preference for the DSR feature would be that we phase out its usage of the NAT map, not extend it further. In particular when this change is currently only needed for a corner-case (in-cluster access to Nodeport service) of a corner-case (pod without SocketLB)

gyutaeb commented on bpf/bpf_lxc.c:
```diff
@@ -112,6 +112,10 @@ lb4_ctx_restore_state(struct __ctx_buff *ctx, struct ct_state *state,
+	__be32 nat_addr = clear ? ctx_load_and_clear_meta(ctx, CB_NAT_ADDR_v4) :
+					ctx_load_meta(ctx, CB_NAT_ADDR_v4);
+	__u16 nat_port = clear ? (__u16)ctx_load_and_clear_meta(ctx, CB_NAT_PORT) :
+					(__u16)ctx_load_meta(ctx, CB_NAT_PORT);
```
@julianwiedmann 
Refactoring based on #43017 has made the code much cleaner :)

The first CT tuple is `pod -> remote node`. Since the remote node address and port are needed for RevNAT, they must be passed to the second CT `pod -> backend`. This works fine for IPv4, but the problem is that there isn't enough space in `cb` to store IPv6 address. Do you have any better ideas for passing the NAT address and port to the second CT?

So, I have left the IPv6 code as the old version.

julianwiedmann replied:
> This works fine for IPv4, but the problem is that there isn't enough space in cb to store IPv6 address. Do you have any better ideas for passing the NAT address and port to the second CT?

A small per-CPU map should do the trick, have a look at how `bpf_lxc` transports the CT lookup result between tailcalls.

gyutaeb commented on bpf/bpf_lxc.c:
```diff
@@ -2055,10 +2131,9 @@ ipv4_policy(struct __ctx_buff *ctx, struct iphdr *ip4, __u32 src_label,
-					   tuple, ipfrag_has_l4_header(fraginfo));
+			ret2 = lb4_rev_nat(ctx, ETH_HLEN, l4_off, ct_state,
+					   ct_state->loopback, tuple,
+					   ipfrag_has_l4_header(fraginfo));
```
> I think we need to do better here, and not have two near-identical calls to `lb4_rev_nat()`.

I've refactored `lb4_rev_nat` to use `ct_state`.

gyutaeb commented on bpf/lib/conntrack.h:
```diff
@@ -280,6 +280,8 @@ ct_lookup_fill_state(struct ct_state *state, const struct ct_entry *entry,
 		state->from_l7lb = entry->from_l7lb;
 		state->from_tunnel = entry->from_tunnel;
+		state->nat_addr = entry->nat_addr;
+		state->nat_port = entry->nat_port;
```
`lb4_rev_nat` needs `state`. So I store nat information to state.

gyutaeb replied:
Same as https://github.com/cilium/cilium/pull/43081

gyutaeb commented on bpf/bpf_lxc.c:
```diff
@@ -59,6 +59,15 @@
 #include "lib/vtep.h"
 #include "lib/subnet.h"
 
+#if defined(ENABLE_DSR)
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, __u32);
+	__type(value, struct dsr_nat_info);
+	__uint(max_entries, 1);
+} cilium_dsr_nat_buffer __section_maps_btf;
+#endif
+
```
> A small per-CPU map should do the trick, have a look at how bpf_lxc transports the CT lookup result between tailcalls.

I used a per-CPU map to store IPv4/IPv6 addresses and ports. It now supports both IPv4/IPv6, and is based on #43017 !

gyutaeb commented on bpf/tests/tc_lxc_lb4_dsr_nodeport.c:
```diff
@@ -0,0 +1,460 @@
+#include "lib/policy.h"
+
+/*
+ * Test: Pod -> Remote NodePort -> Local backend
+ * - Client pod sends packet to remote node's NodePort
+ * - LB selects a backend on the local node (same node as client)
+ * - Packet should be DNATed to local backend IP and port
+ * - Connection tracking entry should be created with
+ * - nat_addr = REMOTE_NODE_IP and nat_port = NODEPORT_PORT
+ */
+PKTGEN("tc", "tc_lxc_dsr_remote_nodeport_local_backend")
+int lxc_dsr_remote_nodeport_local_backend_pktgen(struct __ctx_buff *ctx)
+{
```
> oh nice! Could you please add some additional BPF tests for the NodePort-LB-in-bpf_lxc case? Including the loopback scenario.

@julianwiedmann
I've addressed all your feedback and added 8 test cases.
- Verified that the local backend is selected when accessing a remote node port, and confirmed that reply works correctly.
- Included test cases for the hairpin mode.
- Covered IPv6 scenarios as well.

julianwiedmann reviewed (commented):
:wave: thank you! For review purposes - could you please split this into at least three patches
1. forward path
2. reply path
3.  tests

And then use the patch description to elaborate a bit more on the implementation details. For example on why we need to touch the backend selection.

One aspect we'll need to understand **extremely** well is how this change will interact with existing connections.

gyutaeb committed:
bpf: lxc: Handle DSR for remote NodePort services on source node

When DSR and PER_PACKET_LB are enabled, connections fail if a client pod
sends a request to a remote node's NodePort service while the server
pod is located on the same node as the client.

The root cause is that the remote node only performs DNAT, setting the
packet's source address to client's node address, leading to a hairpin
problem. Consequently, the originating node cannot perform the necessary
REV NAT for the reply packets.

To resolve this, remote NodePort service requests are now handled on the
source node when DSR is enabled, similar to the behavior of socket-level
load balancing.

Implementation details:

- Add lb4/6_lookup_wildcard_nodeport_service(): When ENABLE_DSR is
  defined and the regular service lookup fails, check if the destination
  is a remote node's IP with a NodePort port range. If so, perform a
  wildcard lookup (address=0) to find the NodePort service.

- Use wildcard key for backend selection: When dsr_internal flag is set,
  set key->address to 0 before calling lb4/6_select_backend_id(). This
  applies to both CT_NEW (new connections) and CT_REPLY (backend
  re-selection for existing connections). This is needed for backend
  selection algorithms that use slot lookup (e.g., Random), which look
  up backend slots via lb4/6_lookup_backend_slot() using the service
  key. Without a wildcard key, the lookup would fail because backend
  slot entries are stored with the wildcard service key, not with the
  remote node's IP.

- Store original destination in CT entry: The original destination
  address and port (remote node IP and NodePort) are stored in
  ct_state_new.nat_addr/nat_port, which will be written to the CT entry
  for use in reply path RevNAT processing.

- Use cilium_dsr_nat_buffer per-CPU map: The NAT info is detected in
  __per_packet_lb_svc_xlate_4/6(), but the CT entry is created after DNAT
  when the original destination info is no longer available in the packet.
  The per-CPU buffer preserves this info across the DNAT operation.

Existing connection handling:

- This change only affects DSR traffic destined to remote node's
  NodePort. The wildcard lookup is triggered when lb4/6_lookup_service()
  fails, but it only processes packets where the destination is a remote
  node IP with a port in the NodePort range. Other traffic that fails
  the regular lookup is unaffected.

Fixes: #41962

Co-authored-by: Siwan Kim <siwan.kim@navercorp.com>
Signed-off-by: Gyutae Bae <gyutae.bae@navercorp.com>

gyutaeb committed:
bpf: lxc: Add RevNAT support for DSR remote NodePort connections

Add reverse NAT support for reply packets of DSR remote NodePort
connections. The forward path stores the original destination address
and port in the CT entry's nat_addr/nat_port fields, which are now
used during reply processing.

Implementation details:

- Extend lb4/6_rev_nat() signature: Add nat_addr and nat_port parameters.
  When nat_port is non-zero, use these values directly for RevNAT.
  When nat_port is zero, fall back to the existing rev_nat_index lookup
  for backward compatibility with existing connections.

- Modify ct_lookup_fill_state(): Copy nat_addr and nat_port from the
  CT entry to ct_state, making them available for reply processing.

- Update ipv4/6_policy(): Check for nat_port in addition to
  rev_nat_index when deciding whether to perform RevNAT. Pass the
  CT entry's NAT information to lb4/6_rev_nat().

- Update nodeport_rev_dnat_ipv4/6(): Adapt to the new lb4/6_rev_nat()
  signature by passing NULL/0 for nat_addr/nat_port (these paths use
  the traditional rev_nat_index lookup).

Existing connection handling:

1. New connections (created after this patch):
   - Forward path stores nat_addr/nat_port in CT entry
   - Reply path uses CT entry's nat_addr/nat_port for RevNAT

2. Existing connections (regular NodePort/DSR traffic):
   - CT entry has nat_addr=0, nat_port=0
   - lb4/6_rev_nat checks nat_port first:
     - If nat_port != 0: use nat_addr/nat_port directly
     - If nat_port == 0: fall back to rev_nat_index lookup
   - This ensures existing connections continue to work

3. Upgrade scenario:
   - Existing connections keep working via rev_nat_index fallback
   - New DSR remote nodeport connections use nat_addr/nat_port
   - No connection disruption during rolling upgrade

Co-authored-by: Siwan Kim <siwan.kim@navercorp.com>
Signed-off-by: Gyutae Bae <gyutae.bae@navercorp.com>

gyutaeb committed:
bpf: tests: Add tests for DSR remote NodePort handling

Add BPF unit tests to verify the DSR remote NodePort functionality
for both IPv4 and IPv6.

Test scenarios for DSR mode (tc_lxc_lb4/6_dsr_nodeport.c):

1. Pod -> Remote NodePort -> Local backend (forward path)
   - Client pod sends packet to remote node's NodePort
   - LB selects a backend on the local node (same node as client)
   - Verifies packet is DNATed to local backend IP and port
   - Verifies CT entry contains correct nat_addr (remote node IP)
     and nat_port (NodePort)

2. Local backend -> Pod (reply path)
   - Backend sends reply packet to client
   - Verifies RevNAT is applied correctly
   - Source IP/port changed to remote node IP and NodePort

3. Pod -> Remote NodePort -> Self (hairpin)
   - Client pod sends packet to remote node's NodePort
   - LB selects the client pod itself as the backend
   - Verifies DNAT to client IP and backend port
   - Verifies SNAT to loopback IP for hairpin flow

4. Hairpin reply
   - Pod replies to loopback IP
   - Verifies RevNAT restores remote node IP and NodePort

5. Existing connection handling (UDP)
   - First packet establishes CT entry via legacy path
   - Second packet should use existing CT entry
   - Verifies wildcard lookup is skipped for existing connections

Test scenarios for Hybrid mode (tc_lxc_lb4/6_hybrid_dsr_nodeport.c):

1. DSR service handling
   - Verifies DSR-enabled service triggers wildcard lookup
   - Packet is DNATed to local backend

2. SNAT service handling
   - Verifies SNAT service does NOT trigger wildcard lookup
   - Packet passes through without DNAT (handled by remote node)

Co-authored-by: Siwan Kim <siwan.kim@navercorp.com>
Signed-off-by: Gyutae Bae <gyutae.bae@navercorp.com>

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1681,6 +1741,29 @@ lb4_lookup_service(struct lb4_key *key, const bool east_west)
+	}
+
+	return NULL;
+#endif
```
```suggestion
#ifdef ENABLE_NODEPORT
	const struct remote_endpoint_info *info;
	__u16 service_port;

	service_port = bpf_ntohs(key.dport);
	if (service_port < NODEPORT_PORT_MIN || service_port > NODEPORT_PORT_MAX)
		return NULL;

	info = lookup_ip4_remote_endpoint(key.address, 0);
	if (info && identity_is_remote_node(info->sec_identity)) {
		key.address = 0;
		return lb4_lookup_service(&key, true);
	}
#endif

	return NULL;
```

gyutaeb replied:
Good catch! Applied.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1681,6 +1741,29 @@ lb4_lookup_service(struct lb4_key *key, const bool east_west)
+		return NULL;
+
+	info = lookup_ip4_remote_endpoint(key.address, 0);
+	if (info && identity_is_remote_node(info->sec_identity)) {
```
Can you have another look at the corresponding SocketLB code here, please? In particular whether the checks for `HOST_ID` and `flag_remote_cluster` are needed are as well.

gyutaeb replied:
I checked the SocketLB code regarding your points:
- `HOST_ID` check: this is not needed. Traffic destined for the local host is already matched by `lb4_lookup_service()` before it reaches the wildcard lookup logic.
- `flag_remote_cluster` check: I added `!flag_remote_cluster`. In a Cluster Mesh environment, we should skip the wildcard lookup for services in remote clusters.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1681,6 +1741,29 @@ lb4_lookup_service(struct lb4_key *key, const bool east_west)
+
+	service_port = bpf_ntohs(key.dport);
+	if (service_port < NODEPORT_PORT_MIN || service_port > NODEPORT_PORT_MAX)
+		return NULL;
```
I appreciate this is how the SocketLB code does it currently. But it's actually inefficient, we should be doing the byte-swap on the `NODEPORT_PORT_*` values instead. I'll send a PR for the SocketLB code.

julianwiedmann replied:
https://github.com/cilium/cilium/pull/43869

gyutaeb replied:
I've checked #43869. I agree that merging #43869 for optimization is good idea, even if `NODEPORT_PORT_*` eventually becomes a load-time config.

I will update my code to align with #43869 once it's ready.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -59,6 +59,15 @@
 #include "lib/vtep.h"
 #include "lib/subnet.h"
 
+#if defined(ENABLE_DSR)
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, __u32);
+	__type(value, struct dsr_nat_info);
+	__uint(max_entries, 1);
+} cilium_dsr_nat_buffer __section_maps_btf;
+#endif /* ENABLE_DSR */
```
I think long-term we might want to unify this with `lb4_ctx_store_state()`, and just pass a whole `ct_state` in the scratch buffer. And no longer rely on `skb->cb`.

gyutaeb replied:
Agreed, that would be a nice cleanup.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -173,6 +182,18 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+			ct_state_new.node_port = 1;
+			ct_state_new.dsr_internal = 1;
+			ct_state_new.nat_addr.p4 = tuple.daddr;
+			ct_state_new.nat_port = tuple.sport;
```
Some of this feels redundant. Why does the LB code suddenly require `src_sec_id`? Can we initialize the `dsr_nat_info` in the scratch buffer straight away, without going through the `ct_state`?

gyutaeb replied:
You're right, thanks for the feedback. I've removed the redundant `src_sec_id` assignment and now initialize `dsr_nat_info` directly in the per-CPU buffer at the wildcard match point, without going through `ct_state`.

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -69,6 +69,11 @@ static __always_inline bool ct_state_is_from_l7lb(const struct ct_state *ct_stat
+	union v6addr nat_addr;
+	__be16 nat_port;
+};
+
```
Shouldn't this live in `nodeport.h` instead? Or even `bpf_lxc.c`?

gyutaeb replied:
Good point! I've moved to `bpf_lxc.c`

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -2015,7 +2070,14 @@ static __always_inline int lb4_local(const void *map, struct __ctx_buff *ctx,
+				backend_id = lb4_select_backend_id(ctx, &wildcard_key, tuple, svc);
+			} else {
+				backend_id = lb4_select_backend_id(ctx, key, tuple, svc);
+			}
```
I'm not sure this is the right approach. Ideally we'd already pass the correct `lb4_key` in, based on the matched service. Note that there's a second call to `lb4_select_backend_id()` further down, which you're currently not processing for wildcard lookups?

So let's take a step back, and look at why `lb4_wildcard_lookup_service()` can't take a pointer, and modify the key? Who is looking at the key afterwards, and would get confused by the 0-address?

gyutaeb replied:
First, passing by value(instead of pointer) in `lb4_wildcard_lookup_service()` was intentional. This is because `lb4_dnat_request()` -> `lb4_xlate()` relies on the original `key->address` to calculate the L4 checksum delta:
```c
sum = csum_diff(&key->address, 4, new_daddr, 4, 0);
```
If we modify `key->address` to 0 in place, the checksum calculation would be incorrect. 

You are right about the `CT_REPLY` path (the second `lb4_select_backend_id` call) missing the wildcard lookup. I will include the fix in the next update once we settle on the approach.

julianwiedmann replied:
Ok so if it's only about `lb4_xlate()`, then let's remove the `key` from that path and just take the needed parts (`daddr` and `dport`) from the CT tuple in `lb4_dnat_request()` instead.

Could you try adding such a change as additional patch to the end of the PR, and let's see how much it simplifies the current logic?

gyutaeb replied:
Sounds good. I've addressed the `key` dependency by removing it in the new commit.
https://github.com/cilium/cilium/pull/41963/changes/c8d766f7cfacdfd64c01c0dc9c66671f31e7b96e

gyutaeb replied:
```
❯ make run_bpf_tests
DOCKER_ARGS="--privileged -v /sys:/sys" RUN_AS_ROOT=1 contrib/scripts/builder.sh \
		env MAKEFLAGS="" \
		make  -C bpf/tests/ run \
			"BPF_TEST=""" \
			"BPF_TEST_DUMP_CTX=""" \
			"LOG_CODEOWNERS=" \
			"JUNIT_PATH=" \
			"V=0"
make: Entering directory '/go/src/github.com/cilium/cilium/bpf/tests'
go test ./bpftest \
	-bpf-test-path /go/src/github.com/cilium/cilium/bpf/tests \
	-json -mod=vendor -vet=all -tags=osusergo \
| tparse ;
┌──────────────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │                  PACKAGE                   │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼────────────────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │ 19.53s  │ github.com/cilium/cilium/bpf/tests/bpftest │  --   │ 632  │  0   │  0    │
└──────────────────────────────────────────────────────────────────────────────────────────────┘
make: Leaving directory '/go/src/github.com/cilium/cilium/bpf/tests'
```

julianwiedmann replied:
Sweet, that looks simpler! I'd say let's put this change at the front of the PR then ("de-couple DNAT operation from LB key" or similar), and then base the rest of the PR on it?

gyutaeb replied:
Applied :)

julianwiedmann commented on bpf/tests/tc_lxc_lb4_dsr_nodeport.c:
```diff
@@ -0,0 +1,460 @@
+#define ENABLE_IPV4
+#define ENABLE_NODEPORT
+#define ENABLE_DSR
+#define ENCAP_IFINDEX		42
```
```suggestion
#define ENABLE_IPV4               1
#define ENABLE_NODEPORT   1
#define ENABLE_DSR                1
#define ENCAP_IFINDEX		42
```

for consistency, in case there's any code paths that use `is_defined()`

gyutaeb replied:
Thanks, Applied.

julianwiedmann commented on bpf/tests/tc_lxc_lb4_dsr_nodeport.c:
```diff
@@ -0,0 +1,460 @@
+
+	status_code = data;
+
+	assert(*status_code == CTX_ACT_OK);
+
+	l2 = data + sizeof(__u32);
+	if ((void *)l2 + sizeof(struct ethhdr) > data_end)
+		test_fatal("l2 out of bounds");
+
+	l3 = (void *)l2 + sizeof(struct ethhdr);
+	if ((void *)l3 + sizeof(struct iphdr) > data_end)
+		test_fatal("l3 out of bounds");
+
+	l4 = (void *)l3 + sizeof(struct iphdr);
+	if ((void *)l4 + sizeof(struct tcphdr) > data_end)
+		test_fatal("l4 out of bounds");
+
+	if (l3->daddr != CLIENT_IP)
+		test_fatal("dst IP hasn't been DNATed to client IP (hairpin)");
+
+	if (l4->dest != BACKEND_PORT)
+		test_fatal("dst port hasn't been DNATed to backend port");
+
+	if (l3->saddr != v4_svc_loopback)
+		test_fatal("src IP hasn't been SNATed to loopback IP for hairpin");
+
+	test_finish();
+}
```
Just for completeness, could you also check for the expected CT entry here?

gyutaeb replied:
Thanks, Applied.

julianwiedmann reviewed (commented):
Thank you! Flushing out some feedback from looking over the forward path. Reply path is still todo ...

I'm still trying to wrap my head around the upgrade scenario. If we have an established connection for a client on `NodeA`, which connects to a NodePort on `NodeB` and has been assigned some DSR backend on `NodeC` - how does that work after upgrade? Why would NodeA select the same backend?

gyutaeb commented on bpf/bpf_lxc.c:
```diff
@@ -173,6 +187,40 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+
+		ret = ct_lazy_lookup4(get_ct_map4(&tmp), &tmp, ctx, fraginfo,
+				      l4_off, CT_EGRESS, SCOPE_FORWARD,
+				      CT_ENTRY_ANY, NULL, &monitor);
+		if (ret < 0)
+			return ret;
+
+		/* If a CT_EGRESS entry exists, it indicates the connection was established via
+		 * the legacy path. Preserve this behavior (skip wildcard lookup) to maintain
+		 * consistency for existing flows. Wildcard lookup is applied only for new connections.
+		 */
+		if (ret == CT_NEW) {
+			svc = lb4_wildcard_lookup_service(key);
+			if (svc) {
+				struct dsr_nat_info nat_info = {};
+				__u32 zero = 0;
+
+				ct_state_new.node_port = 1;
+				ct_state_new.dsr_internal = 1;
+
+				nat_info.nat_addr.p4 = tuple.daddr;
+				nat_info.nat_port = tuple.sport;
+				map_update_elem(&cilium_dsr_nat_buffer, &zero, &nat_info, 0);
+			}
+		}
+	}
```
> How about checking whether the `CT_EGRESS` entry for the original connection (without DNAT) exists? This tells us that a previous packet has passed through the `from-container` path. There's some helpers in `nodeport.h` for such "raw" CT lookups, for inspiration.

I've updated the code to preserve existing connections by checking the CT_EGRESS entry, as you suggested.

julianwiedmann replied:
Ah sorry, I misremembered which file has the helpers I suggested for inspiration - it's [conntrack.h](https://github.com/cilium/cilium/blob/6e3657efc0f648d9b1a3f86b842a03c26e485cd1/bpf/lib/conntrack.h#L1201) of course. We should avoid to do a "proper" CT lookup here, as it will potentially impact the statistics & CT state. And has much higher impact on BPF complexity.

So let's do a simple "does this exact entry exist in the map"-style check, as for instance the linked `ct_has_loopback_egress_entry6().`

gyutaeb replied:
Thanks for the pointer to `conntrack.h`. I've updated the code to use `ct_has_egress_entry4()` as suggested.

gyutaeb commented on bpf/tests/tc_lxc_lb4_dsr_nodeport.c:
```diff
@@ -0,0 +1,686 @@
+}
+
+/*
+ * Test: Existing connection - second UDP packet (with NodePort service added)
+ * - Same UDP connection as previous test
+ * - NodePort service is NOW added
+ * - Packet should still NOT be DNATed because CT entry already exists
+ * - This verifies the CT check prevents wildcard lookup for existing connections
+ */
+PKTGEN("tc", "tc_lxc_dsr_existing_conn_udp_second")
+int lxc_dsr_existing_conn_udp_second_pktgen(struct __ctx_buff *ctx)
+{
+	struct pktgen builder;
+	struct udphdr *l4;
+	void *data;
+
+	pktgen__init(&builder, ctx);
+
+	/* Same packet as first test */
+	l4 = pktgen__push_ipv4_udp_packet(&builder,
+					  (__u8 *)client_mac, (__u8 *)node_mac,
+					  CLIENT_IP, REMOTE_NODE_IP,
+					  CLIENT_PORT, NODEPORT_PORT_UDP);
+	if (!l4)
+		return TEST_ERROR;
+
+	data = pktgen__push_data(&builder, default_data, sizeof(default_data));
+	if (!data)
+		return TEST_ERROR;
+
+	pktgen__finish(&builder);
+
+	return 0;
+}
```
> Please also add a test for this - let's say we send one UDP packet to the remote node, then add the wildcard entry, then send another UDP packet.

I also added a test case for existing connection, using UDP.

gyutaeb reviewed (commented):
@julianwiedmann 
Thanks again for the detailed review! I've gone through all the comments and applied the fixes. I'll wait for your input on the remaining open points.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -976,6 +992,28 @@ lb6_lookup_service(struct lb6_key *key, const bool east_west)
 }
 
+static __always_inline const struct lb6_service *
+lb6_wildcard_lookup_service(struct lb6_key key __maybe_unused)
```
```suggestion
lb6_lookup_wildcard_nodeport_service(struct lb6_key key __maybe_unused)
```

Or something similar. Just to emphasize that we're looking for node IPs, and to differentiate from the *other* wildcard path we have in the LB code (matching a special wildcard SVC entry for `LoadBalancer` services)

gyutaeb replied:
Thanks. Applied!

julianwiedmann reviewed (changes requested):
Thank you, glad to see this moving forward nicely! I added some more comments below. Will peek at the reply path next.

One further thought to consider is why we would only apply this behavior when DSR is enabled - in particular because `ENABLE_DSR` doesn't actually mean that *all* services use DSR (they can still use plain SNAT, when DSR is either in hybrid mode or the service has an opt-out annotation). So I'd be fine to focus on DSR for now, but realistically I think we want to open this up to all `ENABLE_NODEPORT` usage. To match what SocketLB is doing, get improved Network-Policy behavior, and cut-out the intermediate hop.

We've had multiple issues in the past with "in-cluster access to remote NodePorts without SocketLB is a mess", **addressing** the whole problem space would be awesome. I'd be happy to invest a few more cycles to make that happen.

gyutaeb committed:
bpf: lb: Decouple DNAT operation from LB key

Instead of passing the lb4_key/lb6_key to lb4_xlate/lb6_xlate for
checksum calculation and port translation, pass the original destination
address and port directly from the CT tuple.

This change:
1. Removes the key parameter from lb4_xlate/lb6_xlate functions
2. Removes the key parameter from lb4_dnat_request/lb6_dnat_request

The CT tuple already contains the same values that were being read from
the key structure:
- tuple->daddr == key->address (original destination)
- tuple->sport == key->dport (reversed port order in CT tuple)

By removing the xlate path's dependency on key, we can now directly
modify key->address = 0 for wildcard lookups without creating a copy,
simplifying the backend selection logic.

Co-authored-by: Siwan Kim <siwan.kim@navercorp.com>
Signed-off-by: Gyutae Bae <gyutae.bae@navercorp.com>

gyutaeb reviewed (commented):
Thanks for sharing the long-term vision. I agree that applying this to all NodePort usage would be ideal. I'll definitely keep that in mind for future contributions.

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -173,6 +187,34 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+		 */
+		if (!ct_has_egress_entry4(get_ct_map4(&tmp), &tmp)) {
+			svc = lb4_lookup_wildcard_nodeport_service(key);
+			if (svc) {
```
```suggestion
			if (svc && nodeport_uses_dsr4(svc)) {
```

Given we only want to handle the DSR path in scope of this PR, I think this constraint is necessary. Otherwise we're unexpectedly also handling non-DSR. All assuming that the wildcard entry in the map is created correctly :).

(this might even be worth a test)

gyutaeb replied:
Agreed :) Making that condition explicit is much clearer. I've updated the code and added a test case as suggested.

julianwiedmann reviewed (changes requested):
Thank you! Just a few minor notes inline. I also had a closer look at the reply path, and that seems all good!

gyutaeb commented on bpf/tests/tc_lxc_lb4_hybrid_dsr_nodeport.c:
```diff
@@ -0,0 +1,294 @@
+#include "common.h"
+#include "pktgen.h"
+
+/* Enable code paths under test */
+#define ENABLE_IPV4		1
+#define ENABLE_NODEPORT		1
+#define ENABLE_DSR		1
+#define ENABLE_DSR_BYUSER	1
+#define ENCAP_IFINDEX		42
```
I've implemented the test in a separate file named `tc_lxc_lb4_hybrid_dsr_nodeport.c` to cover the hybrid scenario. The test verifies that DSR services correctly trigger the wildcard lookup and DNAT, while non-DSR services do not.

gyutaeb reviewed (commented):
@julianwiedmann 
Thanks for the review! I've addressed all the feedback. I also reorganized the commits and updated the commit messages accordingly.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1305,6 +1328,9 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 		if (backend_id == 0) {
+			/* No CT entry has been found, so select a svc endpoint */
+			if (state->node_port && state->dsr_internal)
+				memset(&key->address, 0, sizeof(key->address));
 			backend_id = lb6_select_backend_id(ctx, key, tuple, svc);
```
Do we still need this? I'd expect that the initial reset in `lb6_lookup_wildcard_nodeport_service()` is sufficient.

(and if we don't need it - can we then also do without setting the `node_port` and `dsr_internal` flags?)

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1351,6 +1377,8 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 				goto no_service;
 
+			if (state->node_port && state->dsr_internal)
+				memset(&key->address, 0, sizeof(key->address));
```
Same here.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -976,6 +992,28 @@ lb6_lookup_service(struct lb6_key *key, const bool east_west)
 }
 
+static __always_inline const struct lb6_service *
+lb6_lookup_wildcard_nodeport_service(struct lb6_key key __maybe_unused)
```
```suggestion
lb6_lookup_wildcard_nodeport_service(struct lb6_key *key __maybe_unused)
```

That was the intention, right?

julianwiedmann reviewed (changes requested):
Thank you! Just a bit of fine-tuning left on the `lb_key` handling, otherwise this looks all good.

gyutaeb commented on bpf/bpf_lxc.c:
```diff
@@ -323,6 +365,34 @@ static __always_inline int __per_packet_lb_svc_xlate_6(void *ctx, struct ipv6hdr
 	 * state in the address.
 	 */
 	svc = lb6_lookup_service(&key, is_defined(ENABLE_NODEPORT));
+#if defined(ENABLE_DSR)
+	if (!svc) {
+		struct ipv6_ct_tuple tmp = tuple;
+
+		/* look up with SCOPE_FORWARD: */
+		__ipv6_ct_tuple_reverse(&tmp);
+
+		/* If a CT_EGRESS entry exists, it indicates the connection was
+		 * established via the legacy path. Preserve this behavior (skip
+		 * wildcard lookup) to maintain consistency for existing flows.
+		 * Wildcard lookup is applied only for new connections.
+		 */
+		if (!ct_has_egress_entry6(get_ct_map6(&tmp), &tmp)) {
+			svc = lb6_lookup_wildcard_nodeport_service(&key);
+			if (svc && !nodeport_uses_dsr6(svc))
+				svc = NULL;
+
+			if (svc) {
+				struct dsr_nat_info nat_info = {};
+				__u32 zero = 0;
+
+				ipv6_addr_copy(&nat_info.nat_addr, &tuple.daddr);
+				nat_info.nat_port = tuple.sport;
+				map_update_elem(&cilium_dsr_nat_buffer, &zero, &nat_info, 0);
+			}
+		}
+	}
+#endif /* ENABLE_DSR */
```
> Do we still need this? I'd expect that the initial reset in > lb6_lookup_wildcard_nodeport_service() is sufficient.
>
> (and if we don't need it - can we then also do without setting the node_port and dsr_internal flags?)

@julianwiedmann 
Good point. I've cleaned up the logic as suggested.

Just one note: explicitly setting `svc = NULL` is required to handle non-DSR(hybrid) cases correctly. Previously, the service lookup succeeded even for non-DSR traffic, but it didn't proceed to `lb6_select_backend_id()`. Now that we rely on `key->address = 0`, we must force `svc = NULL` to prevent `lb6_local()` and `lb6_select_backend_id()` from being triggered incorrectly.

I agree that this structure is cleaner and more correct.

julianwiedmann replied:
yeah, makes sense. We could hide that DSR check inside `lb6_lookup_wildcard_nodeport_service()` (and return NULL from there), but that's not much prettier. And it's only temporary anyway (hopefully).

gyutaeb commented on bpf/bpf_lxc.c:
```diff
@@ -173,6 +187,34 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+		 * wildcard lookup) to maintain consistency for existing flows.
+		 * Wildcard lookup is applied only for new connections.
+		 */
+		if (!ct_has_egress_entry4(get_ct_map4(&tmp), &tmp)) {
+			svc = lb4_lookup_wildcard_nodeport_service(&key);
+			if (svc && !nodeport_uses_dsr4(svc))
+				svc = NULL;
```
Same here.

julianwiedmann reviewed (approved):
Looks great now, thanks for pushing this through!

Fingers crossed for CI ...

julianwiedmann commented on bpf/bpf_lxc.c:
```diff
@@ -173,6 +187,34 @@ static __always_inline int __per_packet_lb_svc_xlate_4(void *ctx, struct iphdr *
+		 */
+		if (!ct_has_egress_entry4(get_ct_map4(&tmp), &tmp)) {
+			svc = lb4_lookup_wildcard_nodeport_service(&key);
+			if (svc && !nodeport_uses_dsr4(svc))
```
From unrelated discussion - we could also check for `SVC_FLAG_NODEPORT` here, to be absolutely sure of what we're handling. Assuming that the controlplane sets things up as expected.

Maybe something to look into when relaxing the DSR requirement.

ldelossa reviewed (approved):
Nice change, while policy changes will be needed to communicate thing, translation at source fits our other models much better 👍.

ldelossa reviewed (approved):
LGTM.
